overlord/ifacestate/apparmorprompting/prompting.go

```diff
@@ -0,0 +1,578 @@
+			return err
+		}
+		return nil
+	})
+
+run_loop:
+	for {
+		logger.Debugf("waiting prompt loop")
+		select {
+		case req, ok := <-listenerReqs(currentListener):
+			if !ok {
+				// Reqs() closed, so either errored or Stop() was called.
+				// In either case, the listener Close() method has already
+				// been called, and the tomb error will be set to the return
+				// value of the Run() call from the previous tracked goroutine.
+				logger.Debugf("listener closed requests channel")
+				break run_loop
+			}
+
+			// XXX: this debug log leaks information about internal activity
+			logger.Debugf("Got from kernel req chan: %v", req)
+			if err := m.handleListenerReq(req); err != nil {
+				// XXX: this log leaks information about internal activity
+				logger.Noticef("Error while handling request: %+v", err)
+			}
+		case <-m.tomb.Dying():
+			logger.Noticef("InterfacesRequestsManager tomb is dying, disconnecting")
+			break run_loop
+		}
+	}
+	return m.disconnect()
+}
+
+func (m *InterfacesRequestsManager) handleListenerReq(req *listener.Request) error {
+	userID := uint32(req.SubjectUID)
+	// TODO: immediately deny if req.SubjectUID() == 0 (root)
+	snap := req.Label // Default to apparmor label, in case process is not a snap
+	tag, err := naming.ParseSecurityTag(req.Label)
+	if err == nil {
+		// the triggering process is not a snap, so treat apparmor label as snap field
+		snap = tag.InstanceName()
+	}
+
+	// TODO: when we support interfaces beyond "home", do a proper selection here
+	iface := "home"
+
+	path := req.Path
+
+	permissions, err := prompting.AbstractPermissionsFromAppArmorPermissions(iface, req.Permission)
+	if err != nil {
+		// XXX: this log leaks information about internal activity
+		logger.Noticef("error while parsing AppArmor permissions: %v", err)
+		response := listener.Response{
+			Allow:      false,
+			Permission: req.Permission,
+		}
+		return requestReply(req, &response)
+	}
+
+	remainingPerms := make([]string, 0, len(permissions))
+	satisfiedPerms := make([]string, 0, len(permissions))
+
+	// we're done with early checks, serious business starts now, and we can
+	// take the lock
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	matchedDenyRule := false
+	for _, perm := range permissions {
+		if yesNo, err := m.rules.IsPathAllowed(userID, snap, iface, path, perm); err == nil {
+			if !yesNo {
+				matchedDenyRule = true
+			} else {
+				satisfiedPerms = append(satisfiedPerms, perm)
+			}
+		} else {
+			if !errors.Is(err, requestrules.ErrNoMatchingRule) {
+				logger.Noticef("error while checking request against existing rules: %v", err)
+			}
+			// No matching rule found
+			remainingPerms = append(remainingPerms, perm)
+		}
+	}
+	if matchedDenyRule {
+		// XXX: this debug log leaks information about internal activity
+		logger.Debugf("request denied by existing rule: %+v", req)
+
+		// At least some permissions were allowed, but others were denied.
```

Comment:
shouldn't we check that len(satisfiedPerms) > 0

Reply:
We don't mind whether `len(satisfiedPerms) > 0` or not, the comment is misleading. I'll adjust it to explicitly mention that there be no allowed permissions.

It's increasingly clear to me that the `listener.Response` system (which has fields `Allow bool` and `Permission any`) is unnecessary and confusing, since the `Deny` value basically throws out the `Permission` and just puts all originally requested permissions into the deny field of the message to the kernel $^1$ . A better system would make this more explicit: reply with any permissions you want to allow, any other permissions will be denied. Then the listener can AND the allowed permissions with the originally requested permissions, and send that back to the kernel.

$.^1$ there's a bit of nuance here, the kernel actually sends both an allow and a deny field: the allow field is permissions allowed by the profile, the deny field is permissions matched by rules with the prompt prefix. So we leave the allow field as-is, and treat the deny field as the requested permissions.